### DISCUSSION TITLE
New Docking Logic

### DISCUSSION THREAD

**Discussion by tsunamayo** [tsunamayo/Starship-EVO#3994] [Dev Asks]
Following the big discussion at #3992 that I have finally digested, here is the upcoming plan for docking.
I am in the middle of it as we already have the new connector dock and a spawning system for new modules.

New features:
- New Docking Connector of a smaller form factor - grid-scalable
- New Docking Controller (should be called Spawn Controller) to configure the game procedural generation.
- Station will use a custom docking port, specific for the station design.
- I will let the community come up with a universal station port if people want to exchange modules.
- New docking connector will transfer fuel/inventory/power (I still need to clarify how we link toward the connector).

Ship to Ship / Ship to station docking
- I will give two new squared ports to better fit player design
- Smaller footprint at 2x2m and 2x1m.
- 2x2m can dock to 2x1m (thanks I forgot who for the idea! Was it ProPeach?). It wont be in the middle but on one side though, to align the fuel pipes.
- I will try to make it extensible as suggested, so it extends up to the ship bounds.
- It will have some 0.25cm pipe/inventory connector on the corners to exchange inventory at the same time.

Old Docking Port
- The old docking port gets nuked 
- BUT I will make a simple Iris door mechanism, to be used like existing door and trap door.
- So if you liked it you can still make a custom dock prefab looking just like the old one, with the size you need. Granted it will be bit less clean, but the old port has crappy collisions so I prefer it this way

I think it was important to give a smaller and standard docking port for ships, with a basic "passe partout" aesthetic.
Of course if you dont like it and dont mind having your ship not compatible with other one dont use it!
You will still be able to teleport to the station - it is mostly for immersion and NPCs ships.
For stations, I think we can leverage on the prefab feature and have each builder come up with a docking design as they see fit.

That's it, tell me if I forgot something!

**Comment by Garrett-C** [comment]
Just one thought. Will there be an alternative for fuel and item transfer? I am sort of thinking about smaller fighters here which are unlikely to have full docking ports.

**Reply by tsunamayo** [reply]
I dont know, small fighters can have small docking port right? You wont really play survival with a small fighter anyway, that will be more for carrier when we get to that.

**Reply by Garrett-C** [reply]
I mean a 2x1m docking port doesn't really feel like it would necessarily make sense on a small fighter. Especially with a built-in door. A lot of people do fighters with opening canopies and that would be odd with a door stuck on the bottom of the fighter.

Arguably having a small docking port without the door that's exclusively for item transfer would be the way to go here. A bonus of this would be that the number of items and fuel a player can transfer would be more flexible without needing to add a load of extra doors on your ship. 

Even if most of the gameplay isn't in fighters they will still need to be refueled on carriers etc so some smaller profile mechanism alongside the main one would be ideal imo.

**Reply by tsunamayo** [reply]
yeah sorry I was talking about just the simple docking connector, not the door of course.

**Reply by Garrett-C** [reply]
Oh the one you added in the last update. If so that's perfect yeah. Perhaps it could also have the extending functionality like the doors if you get it working for them. I mean if not a piston and a hose would also do the trick for players.

**Reply by tsunamayo** [reply]
Yeah right, I need to think about piping and children entity. It would make sense to have them pass fuel / inventory.

**Comment by Xerma** [comment]
Super late to the discussion but will docking ship to ship/station be able to transfer stats to the parent? This could be an option in the docking/ship computer ( [x] Allow stat transfer to docked entity). Of course my reasoning is for my beloved Lucrehulk/Core Ship from Star Wars (the core ship/sphere can dock/undock from the ring to deploy to planets) it *could* make for interesting gameplay choices but I'll be honest, I don't see it as terribly necessary. Either way just curious so have some WIP pics 😄 
![20211208160813_1](https://user-images.githubusercontent.com/32974671/145665646-d7678676-18b4-41cc-80e1-4b941ebba70a.jpg)
![20210615114842_1](https://user-images.githubusercontent.com/32974671/145665662-3dc2fdab-7546-4302-9f3f-0b5f8f149f53.jpg)
